File: Q10/cpp/CorporationSalary.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class CorporationSalary
{
public:
    long long totalSalary(vector<string> relations)
    {
        vector<long long> salary = vector<long long>(relations.size(),0);
        
        while (true)
        {
            for(int i =0 ; i<relations.size();i++)
            {
                //급여가 계산된 상태면 더이상 계산할 필요없음.
                if(salary[i] == 0)
                {
                    bool bCheck = false;
                    for(int l = 0 ; l<relations[i].size();l++)
                    {
                        if(relations[i].substr(l,1) == "Y")
                        {
                            //아직 부하급여가 계산이 안되었다면 계산할 필요없음.
                            if(salary[l] == 0)
                            bCheck = true;
                        }
                    }

                    if(!bCheck)
                    {
                        //급여계산시작.
                        int sal = 0 ;
                        bool bFlag = false;
                        for( int l = 0; l<relations[i].size();l++)
                        {
                            if(relations[i].substr(l,1) == "Y")
                            {
                                bFlag = true;
                                sal += salary[l];
                            }
                        }
                        if(!bFlag)
                            sal = 1;
                        salary[i] = sal;
                    }
                }
            }

            //탈출조건 계산
            int condition = 1;
            vector<long long>::iterator it;
            for(it = salary.begin(); it != salary.end(); it++)
                condition *= *it;

            //디버깅
            vector<long long>::iterator it_;
            for(it_ = salary.begin(); it_ != salary.end(); it_++)
                cout<<*it_<<" ";
            cout<<"==============================="<<endl;

            if(condition != 0)
                break;
        }
        

        int sum = 0;
        vector<long long>::iterator it;
        for(it = salary.begin(); it != salary.end(); it++)
            sum += *it;
        return sum;
    }

};

int main()
{
    vector<string> relations_1 = {"N"};//1
    vector<string> relations_2 = {"NNYN","NNYN","NNNN","NYYN"};//5
    vector<string> relations_3 = {"NNNNNN","YNYNNY","YNNNNY","NNNNNN","YNYNNN","YNNYNN"};//17
    vector<string> relations_4 = {"NYNNYN","NNNNNN","NNNNNN","NNYNNN","NNNNNN","NNNYYN"};//8
    vector<string> relations_5 = {"NNNN","NNNN","NNNN","NNNN"}; //4

    cout<<CorporationSalary().totalSalary(relations_1)<<endl;
    cout<<CorporationSalary().totalSalary(relations_2)<<endl;
    cout<<CorporationSalary().totalSalary(relations_3)<<endl;
    cout<<CorporationSalary().totalSalary(relations_4)<<endl;
    cout<<CorporationSalary().totalSalary(relations_5)<<endl;
    return 0;
}

File: Q10/python/CorporationSalary.py
class CorporationSalary:
    def __init__(self):
        pass

    def totalSalary(self, relations : list):
        salary = [0 for i in range(0,relations.__len__())]
        i = 0
        while True:
            for i in range(0,relations.__len__()):
                #급여가 계산된 상태면 더이상 계산할 필요 없음.
                if salary[i] == 0:
                    bCheck : bool = False
                    for l in range(0,relations[i].__len__()):
                        if relations[i][l] == 'Y':
                            #아직 부하 급여가 계산이 안되었다면 계산할 필요 없음.
                            if salary[l] == 0:
                                bCheck = True

                    if bCheck is False:
                        sal : int = 0 # 급여 계산 시작
                        bFlag : bool = False
                        l = 0
                        for l in range(0,relations[i].__len__()):
                            if relations[i][l] == 'Y':
                                bFlag = True
                                sal = sal + salary[l]

                        if bFlag is False:
                            sal = 1
                        salary[i] = sal

            #탈출 조건 계산
            condition : int = 1
            for n in range(0,salary.__len__()):
                condition = condition * salary[n]

            if condition != 0:
                break

        sum : int = 0
        for m in range(0,salary.__len__()):
            sum = sum + salary[m]
        return sum

relations_1 = ["N"]
relations_2 = ["NNYN","NNYN","NNNN","NYYN"]
relations_3 = ["NNNNNN","YNYNNY","YNNNNY","NNNNNN","YNYNNN","YNNYNN"]
relations_4 = ["NYNNYN","NNNNNN","NNNNNN","NNYNNN","NNNNNN","NNNYYN"]
relations_5 = ["NNNN","NNNN","NNNN","NNNN"]

print(CorporationSalary().totalSalary(relations_1))
print(CorporationSalary().totalSalary(relations_2))
print(CorporationSalary().totalSalary(relations_3))
print(CorporationSalary().totalSalary(relations_4))
print(CorporationSalary().totalSalary(relations_5))